src/vm/internal/compiler/lex.cc

```diff
@@ -841,10 +846,79 @@ static int skip_to(const char *token, const char *atoken) {
   }
 }
 
+void init_include_path() {
+  push_malloced_string(add_slash(current_file));        // does master has an include path?
+  svalue_t *ret = apply_master_ob(APPLY_GET_INCLUDE_PATH, 1);
+
+  if (!ret || ret == (svalue_t *)-1) {                  // either no or no master yet
+      return;                                           // just use the runtime configuration
+  }
+  if (ret->type != T_ARRAY) {                           // illegal return value
+      debug_message("'master::get_include_path' must return 'string *'\n");
+      return;                                           // we still have the runtime configuration
+  }
+  array_t *arr  = ret->u.arr;
+  int size = arr->size;
+
+  if(!size) {                                           // empty path?
+      debug_message("got empty include path for 'master::get_include_path(%s)'\n", current_file);
+      return;                                           // we still have the runtime configuration
+  }
+  char **path = static_cast<char **>(DMALLOC(sizeof(char*) * size, TAG_COMPILER, "compiler:init_include_path"));
+
+  // check elements and build working copy
+  int i,                    // index into returned array
+      j,                    // index into dynamic path array
+      k;                    // index into static path array
+  for (i = j = 0; i < arr->size;) {                     // can't use size since it might change
+    if (arr->item[i].type != T_STRING) {                // wrong type of element
+      debug_message("'master::get_include_path(%s)' must return 'string *'\n", current_file);
+      goto init_include_path_cleanup;                   // clean exit
+    }
+
+    const char *elem;
+    if(!strcmp(elem = arr->item[i].u.string, ":DEFAULT:")) { // replace with runtime configuration
+      size += inc_list_size - 1;                        // get additional space
+      path = static_cast<char **>(DREALLOC(path, sizeof(char *) * size, TAG_COMPILER, "compiler:init_include_path"));
+      for (k = 0; k < inc_list_size;) {                 // and copy runtime configuration
+        path[j++] = make_shared_string(inc_list[k++]);
+      }
+    } else if (!legal_path(elem)) {                     // illegal value
+      debug_message("'master::get_include_path(%s)' must give paths without any '..'\n", current_file);
+      goto init_include_path_cleanup;                   // clean exit
+    } else {
+      path[j++] = make_shared_string(elem);             // valid directory
```

Comment:
Currently inc_list is a ```char**``` (as soon as I may compile the driver myself I'll look to change the whole thing to use ```std::string``` instead) since inc_list points to malloced strings those strings ned to be freed, This is independent from the shared strings used later and, as I unstand the shared string logic, the shared strings actualy will be allocated only once:
when the master itself is compiled…
Since all those strings are part of the code of the master they will be allocated as shared strings upon compiling the master and later when used ```make_shared_string()``` just retirns a reference to the allready stored string without actualy "making a new shared string"…